### ISSUE TITLE
Ability to wrap adjacent text styles in a wrapping element/container

### ISSUE THREAD

**Issue by andrew-ireland** [jacksleight/statamic-bard-texstyle#7]
Hi Jack, first of all can I say congratulations on a phenomenal plugin — really well thought out, and a pleasure to use!

I've been toying around with it recently, and had an idea which currently isn't possible, but could be extremely powerful.

Long story short, I've been trying to think of a good way to handle 'Text Columns' using Statamic Bard Textstyle, rather than just adding another set to my Bard field (better editing & visual experience, see visual attached). Unfortunately, due to CSS limitations, the best way to handle columns, is within a dedicated container, with column elements within.

It would be super powerful from a styling perspective on the frontend, if when setting a style there was an option like 'container', or 'wrap' which could be set to true (false by default). I.e.

```
        'two-column' => [
            'type'   => 'paragraph',
            'name'   => 'Two Column Grid',
            'ident'  => '||',
            'class'  => 'o-column',
            'cp_css' => 'display: inline-block; vertical-align: top; width: 50%; padding: 1em;',
            'wrapper' => 'true',  // proposed setting
        ],
```

Basically, this would wrap any elements previous or adjacent with the same selected class together, so styling could be applied directly on the wrapper element. I tried a few methods directly in Bard Mutator to achieve this, but my PHP isn't good enough to hack it together unfortunately! Thought I'd bring the idea to your attention, in case you or anyone else saw the value in it!

Cheers,
Andrew

<img width="922" alt="Screen Shot 2022-04-09 at 8 54 31 pm" src="https://user-images.githubusercontent.com/18650897/162569049-13a63019-2769-4411-a9cc-72eef6481fab.png">

**Comment by jacksleight** [comment]
Hey Andrew, thanks for the feedback, it's always nice to hear people have found this useful :-) .

Some kind of column functionality for Bard is a great idea, although I'm not sure implementing it as proposed would be the best approach, at least not without things getting very hacky.

ProseMirror/Bard needs an actual node to exist in the data in order to render a container element that wraps multiple other nodes, so simply flagging the paragraphs doesn't give us the hooks we need to render the right HTML unfortunately. This is why it's also not possible with Bard Mutator (at the moment). 

The more "ProseMirrory" way to do this would be to introduce a dedicated columns wrapper node, which could be added with a separate editor button. The advantage of that is the structure of the data would be 1:1 with the output HTML, and you could place any content you like inside the columns container (headings, lists etc.) not just paragraphs with a specific style. 

I'm now trying to think if there are other non-column use cases for this kind of wrapper element functionality...

I will definitely take a look at this, I think it'd be a great addition to Texstyle. 👍

**Comment by andrew-ireland** [comment]
Thanks for your reply Jack, and considering the idea. I agree with your thinking regarding the separate editor button; makes sense based on what you've described with how ProseMirror is rendering things under the hood. 

Take my implementation concept (with the wrapper setting) with a massive grain of salt — that was just the best way I could conceive to communicate the idea at the time! 

Excited to see what you come up with down the track. If you want any assistance on the UI/UX side I'm happy to help where I can 🙏
